core/stop/residual_norm_reduction.hpp

```diff
@@ -68,16 +69,14 @@ class ResidualNormReduction
 public:
     using Vector = matrix::Dense<ValueType>;
 
-    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
-    {
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory){
```

Comment:
Different version of clang-format?

Reply:
Actually, I believe it changed the style because I removed the `;` that was down there. The reason was that the compiler I'm using was giving some warnings because due to the macro this `;` was ignored.

Reply:
I know CUDA does it for me in some places. Not sure if we should ignore these warnings or not.
`;` is defined as a null statement, and as such can appear anywhere where a statement can and should be ignored by the compiler. The problem with macros is that they can either define a complete statement (as in `#define F f() {}`), or just a prefix which has to be followed by a statement or used in an expression `#define F f()`. Using the former without a semicolon can cause confusion as it is not clear if code like `F G` means that `G` is used to complete the definition generated by `F`, or if they are two separate macros. On the other hand, always following a macro with a semicolon (even when it is not needed) clarifies this: `F; G;` are two separate statement and `F G;` signifies that `F` modifies the definition of `G`.

The question is what is more important: making clear the meaning of the macro, or making sure there are no false-positive compiler warnings.

Reply:
One option could be to declare a function after macros that define a full statement. The following code compile without warnings with both `clang++ -Wall -Wextra -pedantic -std=c++11` and `g++ -Wall -Wextra -pedantic -std=c++11`:

```c++
#define F void f() {} void __gko_macro_terminator__()
#define G void g() {} void __gko_macro_terminator__()
#define S struct S { int x; } s; void __gko_macro_terminator__()

S;
F;
G;

int main() {
    f();
    auto gx = s.x;
    S;
    auto lx = (s.x = 0);
    return gx + lx;
}
```

Reply:
This option is interesting but for some macros such as the `GKO_ENABLE_CRITERION_FACTORY` we risk polluting the class's public methods declaration (as the end of the macro has to finish with a public declaration).

For now I simply put back the `;` as both in terms of codying style and uniformity it does feel better. Too bad about the warnings, but anyway looking at the compilation output with `-Wall Wextra -pedantic` we actually have this problem in a **lot** of places. I do not know why my compiler was explicitly telling me about the ones from this file. We also have a ton of unused parameters warnings. I will maybe create an issue with this whole discussion and someone can eventually do that later. This could also be a good "beginner" issue.

With GCC 8.2.0
```bash
$ make clean
$ make 2> warnings                                     # without CUDA
$ cat warnings | grep -e "\[-[Wp]" | wc -l       # try to get only the [-W or [-p header for warnings
  25447
$ cat warnings | grep "unused-parameter" | wc -l
  15226
$ cat warnings | grep "extra ‘;’" | wc -l
  9301
```

Reply:
I created issue #174 relating to that discussion. Feel free to contribute there.